app/src/main/java/org/mozilla/tv/firefox/webrender/cursor/CursorModel.kt

```diff
@@ -0,0 +1,352 @@
+                val motionEvent = maybeToMotionEvent(event)
+                HandledKeypress(wasConsumed = motionEvent != null, simulatedTouch = motionEvent)
+            }
+            else -> HandledKeypress(wasConsumed = false, simulatedTouch = null)
+        }
+    }
+
+    /**
+     * @returns true if the event is consumed
+     */
+    private fun directionKeyPress(event: KeyEvent): Boolean {
+        if (!isCursorActive.blockingFirst()) {
+            return false
+        }
+        require(DirectionHelper.KEY_CODES.contains(event.keyCode)) {
+            "Invalid key event passed to CursorController#directionKeyPress: $event"
+        }
+
+        val direction = when (event.keyCode) {
+            KeyEvent.KEYCODE_DPAD_UP -> Direction.UP
+            KeyEvent.KEYCODE_DPAD_DOWN -> Direction.DOWN
+            KeyEvent.KEYCODE_DPAD_LEFT -> Direction.LEFT
+            KeyEvent.KEYCODE_DPAD_RIGHT -> Direction.RIGHT
+            else -> return false
+        }
+        when (event.action) {
+            KeyEvent.ACTION_UP -> directionKeysPressed -= direction
+            KeyEvent.ACTION_DOWN -> {
+                directionKeysPressed += direction
+                pushCursorEvent(direction)
+            }
+            else -> return false
+        }
+
+        _isCursorMoving.onNext(!directionKeysPressed.isEmpty())
+
+        return true
+    }
+
+    private fun pushCursorEvent(direction: Direction) {
+        val edgeNearCursor = getEdgeOfScreenNearCursor()
+        val couldScroll = edgeNearCursor?.let { webViewCouldScrollInDirectionProvider(it) }
+
+        val cursorMovedToEdgeOfScreen = edgeNearCursor == direction
+        val endOfDomContentReached = couldScroll == false
+
+        val event = if (cursorMovedToEdgeOfScreen && endOfDomContentReached) {
+            CursorEvent.ScrolledToEdge(direction)
+        } else {
+            CursorEvent.CursorMoved(direction)
+        }
+
+        _cursorMovedEvents.onNext(event)
+    }
+
+    /**
+     * @return returns a MotionEvent if the event is consumed, null otherwise
+     */
+    @SuppressLint("Recycle")
+    @CheckResult(suggest = "Recycle MotionEvent after use")
+    private fun maybeToMotionEvent(event: KeyEvent): MotionEvent? {
+        if (!isCursorActive.blockingFirst()) {
+            return null
+        }
+        if (event.keyCode != KeyEvent.KEYCODE_DPAD_CENTER && event.keyCode != KeyEvent.KEYCODE_ENTER) {
+            return null
+        }
+
+        val motionEvent = MotionEvent.obtain(event.downTime, event.eventTime, event.action,
+                lastKnownCursorPos.x, lastKnownCursorPos.y, 0)
+        return when (event.action) {
+            KeyEvent.ACTION_UP -> {
+                _isSelectPressed.onNext(false)
+                motionEvent
+            }
+            KeyEvent.ACTION_DOWN -> {
+                _isSelectPressed.onNext(true)
+                motionEvent
+            }
+            else -> null
```

Comment:
The motionEvent doesn't get recycled in this case: we should avoid creating it in this case.